Bricks/net/http/headers/headers.h

```diff
@@ -0,0 +1,309 @@
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef BRICKS_NET_HTTP_HEADERS_HEADERS_H
+#define BRICKS_NET_HTTP_HEADERS_HEADERS_H
+
+#include "../../../port.h"
+
+#include <algorithm>
+#include <cassert>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "../../exceptions.h"
+
+#include "../../../strings/split.h"
+
+namespace current {
+namespace net {
+namespace http {
+
+struct HeaderNotFoundException : Exception {
+  explicit HeaderNotFoundException(const std::string& h) : Exception("Header not found: `" + h + "`.") {}
+};
+
+struct CookieIsNotYourRegularHeader : Exception {
+  using Exception::Exception;
+};
+
+// A generic implementation for HTTP headers.
+// * Shared for client- and server-side use (set by the user or set by an internal HTTP response parser).
+// * Preserves the order in which headers have been added.
+// * Supports lookup by key as well, in case-insensitive way.
+// * Concatenates everything but cookies via ", ", stores cookies as a map.
+// * Preserves case, but allows case-insensitive iteration and lookup. The latter also equalized '-' and '_'.
+struct Header final {
+  const std::string header;  // Immutable as map key, even for a mutable `Header`.
+  std::string value;
+
+  Header() = delete;  // And `header`, the name of the header as string, has to be initalized explicitly.
+  explicit Header(const std::string& header) : header(header) { ThrowIfHeaderIsCookie(header); }
+
+  Header(const std::string& header, const std::string& value) : header(header), value(value) {
+    ThrowIfHeaderIsCookie(header);
+  }
+
+  Header(const Header&) = default;
+  Header(Header&&) = default;
+  Header& operator=(const Header&) = default;
+  Header& operator=(Header&&) = default;
+
+  Header& operator=(const std::string& rhs) {
+    value = rhs;
+    return *this;
+  }
+
+  // Case-insensitive and {'-'/'_'}-insensitive comparison.
+  struct KeyComparator {
+    char Canonical(char c) const {
+      if (c == '-') {
+        return '_';
+      } else if (c >= 'A' && c <= 'Z') {
+        return c + 'a' - 'A';
+      } else {
+        return c;
+      }
+    }
+    bool operator()(const std::string& lhs, const std::string& rhs) const {
+      const size_t lhs_length = lhs.length();
+      const size_t rhs_length = rhs.length();
+      const size_t min_length = std::min(lhs_length, rhs_length);
+      auto lhs_it = lhs.begin();
+      auto rhs_it = rhs.begin();
+      for (size_t i = 0; i < min_length; ++i, ++lhs_it, ++rhs_it) {
+        const char a = Canonical(*lhs_it);
+        const char b = Canonical(*rhs_it);
+        if (a < b) {
+          return true;
+        } else if (b < a) {
+          return false;
+        }
+      }
+      return lhs_length < rhs_length;
+    }
+  };
+
+  // Make sure `Cookies` are never accessed as regular headers.
+  static bool IsCookieHeader(const std::string& header) {
+    const char* kCookieHeaderName = "Cookie";
```

Comment:
It's set in two places where cookies are sent, not received. Let's sync up by voice before I merge it in, ready in 15.